In [34]:
import re
import pandas as pd
import numpy as np
import matplotlib.pyplot as pl
from sklearn.preprocessing import LabelEncoder

In [3]:
df = pd.read_csv("data/raw_analyst_ratings_csv/raw_analyst_ratings.csv")

In [28]:
df.head()

Unnamed: 0                                           headline  \
0           0            Stocks That Hit 52-Week Highs On Friday   
1           1         Stocks That Hit 52-Week Highs On Wednesday   
2           2                      71 Biggest Movers From Friday   
3           3       46 Stocks Moving In Friday's Mid-Day Session   
4           4  B of A Securities Maintains Neutral on Agilent...   

                                                 url          publisher  \
0  https://www.benzinga.com/news/20/06/16190091/s...  Benzinga Insights   
1  https://www.benzinga.com/news/20/06/16170189/s...  Benzinga Insights   
2  https://www.benzinga.com/news/20/05/16103463/7...         Lisa Levin   
3  https://www.benzinga.com/news/20/05/16095921/4...         Lisa Levin   
4  https://www.benzinga.com/news/20/05/16095304/b...         Vick Meyer   

                        date stock  
0  2020-06-05 10:30:54-04:00     A  
1  2020-06-03 10:45:20-04:00     A  
2  2020-05-26 04:30:07-04:00     A  
3  2020-05-22 12:45:06-04:00     A  
4  2020-05-22 11:38:59-04:00     A

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1407328 entries, 0 to 1407327
Data columns (total 6 columns):
 #   Column      Non-Null Count    Dtype 
---  ------      --------------    ----- 
 0   Unnamed: 0  1407328 non-null  int64 
 1   headline    1407328 non-null  object
 2   url         1407328 non-null  object
 3   publisher   1407328 non-null  object
 4   date        1407328 non-null  object
 5   stock       1407328 non-null  object
dtypes: int64(1), object(5)
memory usage: 64.4+ MB


This dataset contains <span style="color:pink">*1,407,328 rows*</span> and <span style="color:pink">*6 columns*</span> with no missing values in any column. Here’s a quick analysis of its structure:

### 1. Columns and Data Types:

* <span style="color:orange">***Unnamed: 0:***</span> Integer index column, <span style="color:red">that uniquely identified the rows </span> likely unnecessary and can be dropped.<br>
* <span style="color:orange">***headline:***</span>  Text data (object), likely representing news headlines. <span style="color:teal">*Sentement texts*</span><br>
* <span style="color:orange">***url:***</span>  Text data (object), contains links to the news articles.<br>
* <span style="color:orange">***publisher:***</span>  Text data (object), represents the source or publisher of the news.<br>
* <span style="color:orange">***date:***</span>  Text data (object), represents the date of the news, which can be converted to datetime for analysis.
* <span style="color:orange">***stock:***</span>  Text data (object), seems to represent stock identifiers or related information.
### 2. Data Characteristics:

* All columns are complete (non-null)<br>
* Most columns are text-heavy (headline, url, publisher, stock) and may require preprocessing for further analysis, especially if this is used in NLP or stock market prediction tasks.<br>
* The date column can be valuable for time-series analysis or trend studies.<br>
### 3. What I planned to do in Next Steps as per the above two results are:

* ***Clean the data***: Drop <span style="color:brown">Unnamed: 0</span> if redundant.
* ***Preprocess the text***: Tokenize or clean <span style="color:brown">headline</span> and <span style="color:brown">publisher</span> for NLP tasks.
* ***Convert <span style="color:brown">date</span> to datetime***: For temporal analysis.
* ***Investigate <span style="color:brown">stock</span>***: Understand its role—does it correspond to stock tickers or categories?

# Preprocessing 

First and for most on preprocessing phase is checking missing values. The below code shows me there is no missing value as well the above information too. 

In [9]:
df[df.isnull().any(axis=1) ]

headline  \
10         30 Stocks Moving in Friday's Pre-Market Session   
11       SVB Leerink Maintains Outperform on Agilent Te...   
12       8 Stocks Moving In Thursday's After-Hours Session   
13       Agilent Technologies shares are trading higher...   
14       Agilent Technologies Q2 Adj. EPS $0.71 Beats $...   
...                                                    ...   
1407323             Top Narrow Based Indexes For August 29   
1407324  Recap: Wednesday's Top Percentage Gainers and ...   
1407325  UPDATE: Oppenheimer Color on China Zenix Auto ...   
1407326  Oppenheimer Initiates China Zenix At Outperfor...   
1407327  China Zenix Auto International Opens For Tradi...   

                                                       url  \
10       https://www.benzinga.com/news/20/05/16092879/3...   
11       https://www.benzinga.com/news/20/05/16092270/s...   
12       https://www.benzinga.com/news/20/05/16089803/8...   
13       https://www.benzinga.com/wiim/20/05/16089218/a...   
14       https://www.benzinga.com/news/earnings/20/05/1...   
...                                                    ...   
1407323  https://www.benzinga.com/news/11/08/1888782/to...   
1407324  https://www.benzinga.com/news/earnings/11/06/1...   
1407325  https://www.benzinga.com/analyst-ratings/analy...   
1407326  https://www.benzinga.com/analyst-ratings/price...   
1407327  https://www.benzinga.com/news/ipos/11/05/10789...   

                       publisher date stock  
10                    Lisa Levin  NaT     A  
11       vishwanath@benzinga.com  NaT     A  
12                  Tyree Gorges  NaT     A  
13             Benzinga Newsdesk  NaT     A  
14             Benzinga Newsdesk  NaT     A  
...                          ...  ...   ...  
1407323            Monica Gerson  NaT    ZX  
1407324             Benjamin Lee  NaT    ZX  
1407325           BenzingaStaffL  NaT    ZX  
1407326                Joe Young  NaT    ZX  
1407327            Allie Wickman  NaT    ZX  

[1351341 rows x 5 columns]

The  next step that I observed there is a duplicated unneccessary index column, that need to be droped. as shown the below code

In [ ]:
# Drop unneccessary duplicated index value as a column lable
df = df.drop(columns=["Unnamed: 0"])

Then ***Date*** columns is stored as ***Object*** datatype while it need to be converted/transformed to ***DateTime***

In [ ]:
#transaform date that are stored as object to date-time data-type
df["date"] = pd.to_datetime(df["date"], errors="coerce")

In [ ]:
#Check data type of date column after transaformation 
df["date"].dtypes

datetime64[ns, UTC-04:00]

In [ ]:
#Check non-null values of a dataframe
df[df['date'].notna()]

headline  \
0                  Stocks That Hit 52-Week Highs On Friday   
1               Stocks That Hit 52-Week Highs On Wednesday   
2                            71 Biggest Movers From Friday   
3             46 Stocks Moving In Friday's Mid-Day Session   
4        B of A Securities Maintains Neutral on Agilent...   
...                                                    ...   
1407266  Chinese Nano-Cap Momentum Stocks Sharply Highe...   
1407267  28 Stocks Moving In Wednesday's Pre-Market Ses...   
1407268  China Zenix Auto International Reports Q3 EPAD...   
1407269                   46 Biggest Movers From Yesterday   
1407270      42 Stocks Moving In Tuesday's Mid-Day Session   

                                                       url          publisher  \
0        https://www.benzinga.com/news/20/06/16190091/s...  Benzinga Insights   
1        https://www.benzinga.com/news/20/06/16170189/s...  Benzinga Insights   
2        https://www.benzinga.com/news/20/05/16103463/7...         Lisa Levin   
3        https://www.benzinga.com/news/20/05/16095921/4...         Lisa Levin   
4        https://www.benzinga.com/news/20/05/16095304/b...         Vick Meyer   
...                                                    ...                ...   
1407266  https://www.benzinga.com/movers/18/01/10994518...      Paul Quintaro   
1407267  https://www.benzinga.com/news/17/12/10878295/2...         Lisa Levin   
1407268  https://www.benzinga.com/news/earnings/17/12/1...      Paul Quintaro   
1407269  https://www.benzinga.com/news/17/11/10788120/4...         Lisa Levin   
1407270  https://www.benzinga.com/news/17/11/10782008/4...         Lisa Levin   

                             date stock  
0       2020-06-05 10:30:54-04:00     A  
1       2020-06-03 10:45:20-04:00     A  
2       2020-05-26 04:30:07-04:00     A  
3       2020-05-22 12:45:06-04:00     A  
4       2020-05-22 11:38:59-04:00     A  
...                           ...   ...  
1407266 2018-01-05 11:47:36-04:00    ZX  
1407267 2017-12-06 09:12:01-04:00    ZX  
1407268 2017-12-06 07:04:31-04:00    ZX  
1407269 2017-11-15 06:04:52-04:00    ZX  
1407270 2017-11-14 13:25:57-04:00    ZX  

[55987 rows x 5 columns]

In [13]:
#Drop invalid datetime value after coversion 
df = df[df['date'].notna()]

In [24]:
# Transform headline or sentiment to lower case, remove numbers and special character
df.loc[:,'headline'] = df['headline'].str.lower().apply(lambda x: re.sub(r'[^a-zA-Z\s]', '', x))


I also believe that ***url*** is not important for analysis, So I decieded to drop the ***url*** as a column label

In [25]:
df = df.drop(columns=["url"])

In [26]:
df.head()

headline          publisher  \
0               stocks that hit week highs on friday  Benzinga Insights   
1            stocks that hit week highs on wednesday  Benzinga Insights   
2                         biggest movers from friday         Lisa Levin   
3            stocks moving in fridays midday session         Lisa Levin   
4  b of a securities maintains neutral on agilent...         Vick Meyer   

                       date stock  
0 2020-06-05 10:30:54-04:00     A  
1 2020-06-03 10:45:20-04:00     A  
2 2020-05-26 04:30:07-04:00     A  
3 2020-05-22 12:45:06-04:00     A  
4 2020-05-22 11:38:59-04:00     A

In [32]:
# To check the unique values of 'publisher' column I can use either
df["publisher"].nunique()
# or
df['publisher'].value_counts()

publisher
Benzinga Newsdesk    14750
Lisa Levin           12408
ETF Professor         4362
Paul Quintaro         4212
Benzinga Newsdesk     3177
                     ...  
Jennifer Lynn            1
Tatro Capital            1
webmaster                1
Thomas Stein             1
Usman Rafi               1
Name: count, Length: 225, dtype: int64

Here we have <span style = 'color:orange'>***255***</span> unique values

In [33]:
df.shape

(55987, 4)

I have an options to use <span style = "color:teal">***one-hot-encoding***</span> for encoding but, in my case, since there are 
255 unique publishers among thetotal of 55987,this might lead to high dimensionality (many columns), which can be inefficient and hard to manage in machine learning models. Therefore, using <span style = "color:teal">***one-hot-encoding***</span> is inefficient

Therefore, <span style = "color:teal">***Label Encoding***</span> recommended for large number of categories as like the data I have

In [35]:
# Initialize the label encoder
encoder = LabelEncoder()

# Apply label encoding to the 'publisher' column
df['publisher'] = encoder.fit_transform(df['publisher'])

In [36]:
df.head()

headline  publisher  \
0               stocks that hit week highs on friday         18   
1            stocks that hit week highs on wednesday         18   
2                         biggest movers from friday        122   
3            stocks moving in fridays midday session        122   
4  b of a securities maintains neutral on agilent...        214   

                       date stock  
0 2020-06-05 10:30:54-04:00     A  
1 2020-06-03 10:45:20-04:00     A  
2 2020-05-26 04:30:07-04:00     A  
3 2020-05-22 12:45:06-04:00     A  
4 2020-05-22 11:38:59-04:00     A

In [ ]:

df['stock'].value_counts()

stock
ZX       10
A        10
AA       10
AAC      10
ZUMZ     10
         ..
ATMP      1
ACCU      1
ZMLP      1
ZIONW     1
SMLL      1
Name: count, Length: 6204, dtype: int64

In [1]:
df.duplicated()

NameError: name 'df' is not defined

In [ ]:
df = df.drop_duplicates()